src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -1879,6 +1885,82 @@ private BoundExpression BindCast(CastExpressionSyntax node, DiagnosticBag diagno
+            GetWellKnownTypeMember(Compilation, WellKnownMember.System_Index__ctor, diagnostics, syntax: node);
+
+            return BindUnaryOperator(node, diagnostics);
+        }
+
+        private BoundExpression BindRangeExpression(RangeExpressionSyntax node, DiagnosticBag diagnostics)
+        {
+            CheckFeatureAvailability(node, MessageID.IDS_FeatureRangeOperator, diagnostics);
+            TypeSymbol rangeType = GetWellKnownType(WellKnownType.System_Range, diagnostics, node);
+
+            if (!rangeType.IsErrorType())
+            {
+                if (node.LeftOperand is null)
+                {
+                    if (node.RightOperand is null)
+                    {
+                        GetWellKnownTypeMember(Compilation, WellKnownMember.System_Range__All, diagnostics, syntax: node);
+                    }
+                    else
+                    {
+                        GetWellKnownTypeMember(Compilation, WellKnownMember.System_Range__ToEnd, diagnostics, syntax: node);
+                    }
+                }
+                else
+                {
+                    if (node.RightOperand is null)
+                    {
+                        GetWellKnownTypeMember(Compilation, WellKnownMember.System_Range__FromStart, diagnostics, syntax: node);
+                    }
+                    else
+                    {
+                        GetWellKnownTypeMember(Compilation, WellKnownMember.System_Range__Create, diagnostics, syntax: node);
+                    }
+                }
+            }
+
+            BoundExpression left = BindRangeExpressionOperand(node.LeftOperand, diagnostics);
+            BoundExpression right = BindRangeExpressionOperand(node.RightOperand, diagnostics);
+
+            if ((left != null && left.Type.IsNullableType()) || (right != null && right.Type.IsNullableType()))
+            {
+                rangeType = GetSpecialType(SpecialType.System_Nullable_T, diagnostics, node).Construct(rangeType);
+            }
+
+            return new BoundRangeExpression(node, left, right, rangeType);
+        }
+
+        private BoundExpression BindRangeExpressionOperand(ExpressionSyntax operand, DiagnosticBag diagnostics)
+        {
+            if (operand is null)
+            {
+                return null;
+            }
+
+            BoundExpression boundOperand = BindValue(operand, diagnostics, BindValueKind.RValue);
+            TypeSymbol indexType = GetWellKnownType(WellKnownType.System_Index, diagnostics, operand);
+
+            if (boundOperand.Type.IsNullableType())
+            {
+                indexType = GetSpecialType(SpecialType.System_Nullable_T, diagnostics, operand).Construct(indexType);
+            }
+
+            HashSet<DiagnosticInfo> notUsed = null;
+            Conversion conversion = this.Conversions.ClassifyImplicitConversionFromExpression(boundOperand, indexType, ref notUsed);
```

Comment:
Should the diagnostics in `notUsed` be reported?

Reply:
`GenerateImplicitConversionError()` should report errors. We need to add tests for more scenarios for operands and their types/ref kinds, but I'll keep that to a later PR.

---
In reply to: [185078426](https://github.com/dotnet/roslyn/pull/26405#discussion_r185078426) [](ancestors = 185078426)

Reply:
Should the diagnostics be added regardless of whether the conversion was valid? And please add a `// PROTOTYPE` comment to test use-site diagnostics.

---
In reply to: [185106036](https://github.com/dotnet/roslyn/pull/26405#discussion_r185106036) [](ancestors = 185106036,185078426)